### PR TITLE
update vcat to mimic Base.vcat and enhance promotion rules of mixed column type

### PR DESCRIPTION
**Pull request by cjprybol**
The current `vcat` operates on Arrays of DataTables `DataTable[]`, while `Base.vcat` utilizes slurping to capture any number of inputs as a tuple or argument to concatenate. This PR changes `vcat` to follow `Base.vcat`'s lead and utilizes a function structure of `vcat(dts::AbstractDataTable...)`. Uses @nalimilan's `@generated` function to implement a new type of `AbstractArray` promotion rule that ideally can be tested here and transfered into Base in the near future https://github.com/JuliaLang/julia/pull/20815. This also removes the prior assumptions that were made about how to join datatables that were out of order or had columns that were only present in some of the arguments and not others.

### CODE REVIEW

File: src/abstractdatatable/abstractdatatable.jl [hunk 1]
```diff
@@ -706,83 +706,96 @@ Base.hcat(dt1::AbstractDataTable, dt2::AbstractDataTable) = hcat!(dt[:, :], dt2)
...
+│ 6   │ 6 │ 6 │
+```
+"""
+Base.vcat(dt::AbstractDataTable) = dt
+function Base.vcat(dts::AbstractDataTable...)
+    isempty(dts) && return DataTable()
+    allheaders = map(names, dts)
+    # don't vcat empty DataTables
+    notempty = find(x -> length(x) > 0, allheaders)
+    uniqueheaders = unique(allheaders[notempty])
```
Comment (nalimilan) [inline src/abstractdatatable/abstractdatatable.jl]: I'm not sure about this: is there a strong reason not to check headers of empty tables? If they don't match, it could indicate a bug in the user code, and I can't find a situation where this would be useful.

Also, it's not correct to return `DataTable()` in that case since that doesn't preserve the headers if only empty tables are passed.

File: src/abstractdatatable/abstractdatatable.jl [hunk 2]
```diff
@@ -706,83 +706,96 @@ Base.hcat(dt1::AbstractDataTable, dt2::AbstractDataTable) = hcat!(dt[:, :], dt2)
...
+    if length(uniqueheaders) > 1
+        unionunique = union(uniqueheaders...)
+        coldiff = setdiff(unionunique, intersect(uniqueheaders...))
+        if !isempty(coldiff)
+            # if any datatables are a full superset of names, skip them
+            filter!(u -> Set(u) != Set(unionunique), uniqueheaders)
+            estrings = Vector{String}(length(uniqueheaders))
+            for (i, u) in enumerate(uniqueheaders)
+                matchingloci = find(h -> u == h, allheaders)
+                headerdiff = filter(x -> !in(x, u), coldiff)
```
Comment (nalimilan) [inline src/abstractdatatable/abstractdatatable.jl]: This is just `setdiff(coldiff, u)`, right?

File: src/abstractdatatable/abstractdatatable.jl [hunk 3]
```diff
@@ -706,83 +706,96 @@ Base.hcat(dt1::AbstractDataTable, dt2::AbstractDataTable) = hcat!(dt[:, :], dt2)
...
+                headerdiff = join(string.(headerdiff), ", ", " and ")
```
Comment (nalimilan) [inline src/abstractdatatable/abstractdatatable.jl]: Avoid reusing variable names for values of different types, this creates a type instability. Same below.

Also `string` isn't needed here an below. So you might be able to put the commands directly inside the string literal.

File: src/abstractdatatable/abstractdatatable.jl [hunk 4]
```diff
@@ -706,83 +706,96 @@ Base.hcat(dt1::AbstractDataTable, dt2::AbstractDataTable) = hcat!(dt[:, :], dt2)
...
+    end
+    if length(uniqueheaders) > 1
+        unionunique = union(uniqueheaders...)
+        coldiff = setdiff(unionunique, intersect(uniqueheaders...))
+        if !isempty(coldiff)
+            # if any datatables are a full superset of names, skip them
+            filter!(u -> Set(u) != Set(unionunique), uniqueheaders)
+            estrings = Vector{String}(length(uniqueheaders))
+            for (i, u) in enumerate(uniqueheaders)
+                matchingloci = find(h -> u == h, allheaders)
```
Comment (nalimilan) [inline src/abstractdatatable/abstractdatatable.jl]: Is "loci" really useful here?

File: src/abstractdatatable/abstractdatatable.jl [hunk 5]
```diff
@@ -706,83 +706,96 @@ Base.hcat(dt1::AbstractDataTable, dt2::AbstractDataTable) = hcat!(dt[:, :], dt2)
...
-        if nonnull_ct[j] < length(dts) && !_isnullable(similars[j])
-            similars[j] = EMPTY_DATA
+    else
+        header = uniqueheaders[1]
+        dts_to_vcat = dts[notempty]
+        cols = Vector{Any}(length(header))
+        for i in 1:length(cols)
+            data = [dt[i] for dt in dts_to_vcat]
+            res = promote_col_type(data...)(mapreduce(length, +, data))
+            cols[i] = copy!(res, vcat(data...))
```
Comment (nalimilan) [inline src/abstractdatatable/abstractdatatable.jl]: Calling `vcat` will create a temporary array. Better repeatedly call `copy!` with a changing offset (see `do` argument).

File: src/groupeddatatable/grouping.jl [hunk 6]
```diff
@@ -193,7 +193,7 @@ combine(map(d -> mean(dropnull(d[:c])), gd))
 """
 function combine(ga::GroupApplied)
     gd, vals = ga.gd, ga.vals
-    valscat = vcat(vals)
+    valscat = vcat(vals...)
```
Comment (nalimilan) [inline src/groupeddatatable/grouping.jl:196]: Can this go into a separate commit?

**Review by nalimilan — COMMENTED** [review]
Thanks, looks mostly good. Please copy the explanations to the commit message (BTW, I think by "slurping" you mean "splatting"?).

It's not completely correct to say "The current vcat operates on Arrays of DataTable", since the vararg form was also supported: what you are doing is that you remove the form taking a vector of tables.

File: src/groupeddatatable/grouping.jl (hunk 6 above)
Comment (cjprybol) [reply src/groupeddatatable/grouping.jl:196]: Different commit but same PR? I can do that 👍

File: src/abstractdatatable/abstractdatatable.jl (hunk 1 above)
Comment (cjprybol) [reply src/abstractdatatable/abstractdatatable.jl]: I think we agree here, but just to clarify
```julia
julia> vcat(DataTable(), DataTable(), DataTable()) #this is correct
0×0 DataTables.DataTable


julia> vcat(DataTable(), DataTable(), DataTable(x=[])) # this should be an error
0×1 DataTables.DataTable


julia> vcat(DataTable(), DataTable(), DataTable(x=[1])) # should also be an error
1×1 DataTables.DataTable
│ Row │ x │
├─────┼───┤
│ 1   │ 1 │
```

**Comment by cjprybol** [comment]
Thanks for the review! I'll get to these edits soon. re: slurping/splatting, I'm not sure if I used the right one. Both descriptions are used [in the docs](http://docs.julialang.org/en/latest/manual/faq.html#The-two-uses-of-the-...-operator:-slurping-and-splatting-1), but I'll say something more explicit in the commit to avoid any ambiguity.

File: src/abstractdatatable/abstractdatatable.jl (hunk 1 above)
Comment (nalimilan) [reply src/abstractdatatable/abstractdatatable.jl]: Yes, that's what I meant.

**Comment by nalimilan** [comment]
Funny, I didn't remember that term. It's fine then.

**Comment by cjprybol** [comment]
That should address all of the comments. I added more tests to assert that the array promotion was working as intended, and a couple tests for vcatting 3 or more datatables to make sure my offset `copy!`s were working as intended. I think I'm finally starting to get squashing/rebasing/amending commits too. Coverage drops seem unrelated and should be addressed by https://github.com/JuliaData/DataTables.jl/pull/31.

**Comment by cjprybol** [comment]
@nalimilan

File: src/abstractdatatable/abstractdatatable.jl [hunk 7]
```diff
@@ -706,83 +706,97 @@ Base.hcat(dt1::AbstractDataTable, dt2::AbstractDataTable) = hcat!(dt[:, :], dt2)
...
-    for j in 1:length(colindex)
-        if nonnull_ct[j] < length(dts) && !_isnullable(similars[j])
-            similars[j] = EMPTY_DATA
+    else
+        header = uniqueheaders[1]
+        cols = Vector{Any}(length(header))
+        for i in 1:length(cols)
+            data = [dt[i] for dt in dts]
+            lens = map(length, data)
+            cols[i] = promote_col_type(data...)(reduce(+, lens))
```
Comment (nalimilan) [inline src/abstractdatatable/abstractdatatable.jl]: Why not just `sum(lens)`? Same below.

File: src/abstractdatatable/abstractdatatable.jl [hunk 8]
```diff
@@ -706,83 +706,97 @@ Base.hcat(dt1::AbstractDataTable, dt2::AbstractDataTable) = hcat!(dt[:, :], dt2)
...
+            copy!(cols[i], data[1])
+            for j in 2:length(data)
+                offset = reduce(+, lens[1:j-1]) + 1
```
Comment (nalimilan) [inline src/abstractdatatable/abstractdatatable.jl]: `offset = lens[1]` for initialization and `offset += lens[j]` after each iteration would be simpler.

File: test/cat.jl [hunk 9]
```diff
@@ -137,15 +116,121 @@ module TestCat
...
+    @testset "vcat mixed coltypes" begin
+        dt = vcat(DataTable([[1]], [:x]), DataTable([[1.0]], [:x]))
+        @test dt == DataTable([[1.0, 1.0]], [:x])
+        @test eltypes(dt)[1] == Float64
+        dt = vcat(DataTable([[1]], [:x]), DataTable([["1"]], [:x]))
+        @test dt == DataTable([[1, "1"]], [:x])
+        @test eltypes(dt)[1] == Any
+        dt = vcat(DataTable([NullableArray([1])], [:x]), DataTable([[1]], [:x]))
+        @test dt == DataTable([NullableArray([1, 1])], [:x])
+        @test eltypes(dt)[1] == Nullable{Int}
```
Comment (nalimilan) [inline test/cat.jl]: Can you rather test the type of the column, to make sure we get a `NullableArray` rather than an `Array{Nullable}`? Same below for `(Nullable)CategoricalArray` vs. Array{CategoricalValue}`.

File: test/grouping.jl [hunk 10]
```diff
@@ -146,8 +146,9 @@ module TestGrouping
     groupby(dt, [:v1, :v2])
 
     dt2 = by(e->1, DataTable(x=Int64[]), :x)
-    @test size(dt2) == (0,1)
+    @test size(dt2) == (0,2)
     @test isequal(sum(dt2[:x]), Nullable(0))
+    @test size(by(e->1, DataTable(x=Int64[1]), :x)) == (1,2)
```
Comment (nalimilan) [inline test/grouping.jl]: Why add this line in particular, and check only its size? It looks out of place in this block. Either remove it or make it more complete in its own block.

**Review by nalimilan — COMMENTED** [review]
Sorry, it's hard for me to keep up with the reviews. Should be good to go after this round.

**Comment by cjprybol** [comment]
No worries! It's not that easy to keep up with the rate at which you do manage to review either. Thanks, as always, for putting in the time to help me improve these PRs.

**Comment by cjprybol** [comment]
Coverage drop doesn't seem related here either. Several functions flagged by Coveralls are handled by https://github.com/JuliaData/DataTables.jl/pull/31

File: test/cat.jl [hunk 11]
```diff
@@ -137,15 +116,119 @@ module TestCat
...
+        @test typeof.(dt.columns) == [Vector{Float64}]
+        dt = vcat(DataTable([[1]], [:x]), DataTable([["1"]], [:x]))
+        @test dt == DataTable([[1, "1"]], [:x])
+        @test typeof.(dt.columns) == [Vector{Any}]
+        dt = vcat(DataTable([NullableArray([1])], [:x]), DataTable([[1]], [:x]))
+        @test dt == DataTable([NullableArray([1, 1])], [:x])
+        @test typeof.(dt.columns) == [NullableVector{Int}]
+        dt = vcat(DataTable([CategoricalArray([1])], [:x]), DataTable([[1]], [:x]))
+        @test dt == DataTable([CategoricalArray([1, 1])], [:x])
+        @test typeof.(dt.columns) == [CategoricalVector{Int, UInt32}]
```
Comment (nalimilan) [inline test/cat.jl]: Leave out the `UInt32` (or replace it with `CategoricalArrays.DefaultRefType`, but probably not worth it).

File: test/grouping.jl [hunk 12]
```diff
@@ -148,7 +148,6 @@ module TestGrouping
     dt2 = by(e->1, DataTable(x=Int64[]), :x)
     @test size(dt2) == (0,2)
     @test isequal(sum(dt2[:x]), Nullable(0))
-    @test size(by(e->1, DataTable(x=Int64[1]), :x)) == (1,2)
```
Comment (nalimilan) [inline test/grouping.jl]: Should put this in the first commit.

**Comment by cjprybol** [comment]
Thanks!
